### PR TITLE
Implementation dependency mapping

### LINKED ISSUE #65: Bindings being generated for non pub methods
Non pub functions are being generated alongside pub functions
example:

```rust
#[ligen(c)]
impl Adder {
    fn string_test(a: String) -> String {
        println!("{:#?}", a);
        a
    }
}
```

has a header entry generated:

```c
#pragma once

#include <stdint.h>

#ifdef __cplusplus
extern "C" {
#endif

typedef struct Struct_Adder {
void* self;
} Adder;

RString Adder_string_test(const char* a);
void Adder_destroy(Adder adder);

#ifdef __cplusplus
}
#endif

```

### PR DESCRIPTION
**Pull request by arrudagates**

### CODE REVIEW

File: core/src/ir/implementation.rs [hunk 1]
```diff
@@ -68,6 +68,44 @@ impl TryFrom<ItemImpl> for Implementation {
...
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                deps.extend(input_deps);
+                if let Some(Type::Compound(ident)) = method.output.clone() {
+                    if !deps.iter().any(|inner| inner.name == ident.name)
+                    // FIXME: Check if type is a dependency
+                        && ident.name != "Self"
```
[the remark this answers is not in the document]
Comment (arrudagates) [reply core/src/ir/implementation.rs]: It already has PartialEq

File: core/src/ir/implementation.rs [hunk 2]
```diff
@@ -8,7 +8,7 @@ use syn::{parse2, ItemImpl};
 pub struct Implementation {
     /// attributes field
     pub attributes: Attributes,
-    /// self_ field
+    /// self_ f  kind: (), is_constant: (), type_: ()kind: (), is_constant: (), type_: () kind: (), is_constant: (), type_: () kind: (), is_constant: (), type_: ()ield
```
Comment (notdanilo) [inline core/src/ir/implementation.rs]: What is this?

File: core/src/ir/implementation.rs [hunk 3]
```diff
@@ -68,14 +68,40 @@ impl TryFrom<ItemImpl> for Implementation {
...
 
+impl Implementation {
+    /// Maps the dependencies in the method signatures
+    pub fn dependencies(&self) -> Vec<Type> {
+        let mut deps: Vec<Type> = vec![];
+        for item in &self.items {
+            if let ImplementationItem::Method(method) = item {
+                method.inputs.clone().into_iter().for_each(|parameter| {
+                    if !deps.iter().any(|typ| typ == &parameter.type_) {
+                        deps.push(parameter.type_);
```
Comment (notdanilo) [inline core/src/ir/implementation.rs:79]: Don't we also need to check against `Self`? It isn't guaranteed that the type will be transformed when this function is called

**Review by notdanilo — CHANGES_REQUESTED** [review]
Few questions.

File: core/src/ir/implementation.rs (hunk 2 above)
Comment (arrudagates) [reply core/src/ir/implementation.rs]: Sorry, i must have clicked the paste button

File: core/src/ir/implementation.rs (hunk 3 above)
Comment (arrudagates) [reply core/src/ir/implementation.rs:79]: You mean `Self` or `self`?

**Review by notdanilo — CHANGES_REQUESTED** [review]

**Review by arrudagates — COMMENTED** [review]
Added test to demonstrate Self

**Review by notdanilo — DISMISSED** [review]
OK

**Review by notdanilo — DISMISSED** [review]
OK

**Review by notdanilo — APPROVED** [review]
OK

**Pull request merged**